### PR TITLE
[expotools] add `et android-build-packages` (replacing `gulp update-exponent-view`)

### LINKED ISSUE #5254: Revisit "update-exponent-view"; maybe move to expotools
- Do we need `update-exponent-view`?
- Should we move it to expotools?
- Anything else to make it easier it work with?

### PR DESCRIPTION
**Pull request by esamelson**
# Why

Resolves #5254 .

Adds a new expotools command, `android-build-packages` that replaces the previous functionality of `gulp update-exponent-view` (which had evolved to building all android packages rather than just expoview).

The old command had become quite unwieldy as new unimodules were added. It cleared all old build artifacts and rebuilt every package from scratch, even if there were no changes. It took ~10 mins to run on my machine, and any errors meant starting over from scratch. Except when adding a new SDK version, we usually never need to rebuild all packages, just a specific few.

The new command can accept a list of packages to rebuild. If the user chooses not to specify, there is an interactive prompt. The script can also intelligently guess which packages need to be rebuilt by looking at git logs in the source and build dirs. This saves a ton of time when there are only a few packages that need to be built.

I also cleaned up and modularized the code for the `updateExpoViewAsync` function, as well as logging more useful output.

Finally, there is better error handling. Gradle errors are logged now, instead of being swallowed. Further, if a package fails to build, it will not prevent the rest of the script from running; instead, the error is logged and instructions are presented to the user at the end of the script.

# Test Plan

Ran the command with `suggested` option on master; this was sufficient to make `et android-shell-app` successfully build a shell app that runs on a device. Still need to test building all packages, ReactAndroid apparently needs to download boost and this internet is too slow for that 😅 

TODO:
- [x] update release instructions
- [x] grep for `gulp update-exponent-view` in other documentation
- [x] test building all packages

### CODE REVIEW

File: tools/expotools/src/commands/AndroidBuildPackages.ts [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+  } else {
+    // gather suggested package data and then show prompts
+    console.log(' 🔍  Gathering data...');
+
+    packagesToBuild = await _getSuggestedPackagesToBuild(packages);
+
+    console.log(' 🕵️   It appears that the following packages need to be rebuilt:');
+    console.log(packagesToBuild);
+
+    const { option } = await inquirer.prompt([{
```
Comment (tsapeta) [inline tools/expotools/src/commands/AndroidBuildPackages.ts]: Does it compile well with TypeScript? Imho you should do this:
```suggestion
    const { option } = await inquirer.prompt<{ option: string }>([{
```

File: tools/expotools/src/commands/AndroidBuildPackages.ts [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+        { value: 'suggested', name: 'Build the suggested packages only' },
+        { value: 'all', name: 'Build all packages' },
+        { value: 'choose', name: 'Choose packages manually' },
+      ],
+    }]);
+
+    if (option === 'all') {
+      packagesToBuild = packages.map(pkg => pkg.name);
+    } else if (option === 'choose') {
+      const result: any = await inquirer.prompt([{
```
Comment (tsapeta) [inline tools/expotools/src/commands/AndroidBuildPackages.ts]:
```suggestion
      const result = await inquirer.prompt<{ packagesToBuild: string[] }>([{
```

File: tools/expotools/src/commands/AndroidBuildPackages.ts [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+    readline.clearLine(process.stdout, 0);
+    readline.cursorTo(process.stdout, 0);
+    process.stdout.write(` ✅  Finished building ${pkg.name}\n`);
+  }
+
+  await _restoreFilesAsync(filesToStash);
+
+  console.log(' 🚚  Copying newly built packages...');
+
+  await spawnAsync('mkdir', ['-p', path.join(ANDROID_DIR, 'maven/com/facebook')])
```
Comment (tsapeta) [inline tools/expotools/src/commands/AndroidBuildPackages.ts]: `fs-extra` has a method that does the same as `mkdir -p` which would look better than `spawnAsync` and works on all platforms:
```suggestion
  await fs.mkdirp(path.join(ANDROID_DIR, 'maven/com/facebook'));
```

File: tools/expotools/src/commands/AndroidBuildPackages.ts [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+  await spawnAsync('mkdir', ['-p', path.join(ANDROID_DIR, 'maven/host/exp/exponent')]);
```
Comment (tsapeta) [inline tools/expotools/src/commands/AndroidBuildPackages.ts]:
```suggestion
  await fs.mkdirp(path.join(ANDROID_DIR, 'maven/host/exp/exponent'));
```

File: tools/expotools/src/commands/AndroidBuildPackages.ts [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+  await spawnAsync('mkdir', ['-p', path.join(ANDROID_DIR, 'maven/org/unimodules')]);
```
Comment (tsapeta) [inline tools/expotools/src/commands/AndroidBuildPackages.ts]:
```suggestion
  await fs.mkdirp(path.join(ANDROID_DIR, 'maven/org/unimodules'));
```

File: tools/expotools/src/commands/AndroidBuildPackages.ts [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+
+  for (const pkg of packages) {
+    await spawnAsync('cp', [
```
Comment (tsapeta) [inline tools/expotools/src/commands/AndroidBuildPackages.ts]: Use `fs.copy` instead

File: tools/expotools/src/commands/AndroidBuildPackages.ts [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+      '-r',
+      path.join(process.env.HOME!, '.m2', 'repository', pkg.buildDirRelative),
+      path.join(ANDROID_DIR, 'maven', pkg.buildDirRelative),
+    ]);
+  }
+
+  // Copy JSC
+  await spawnAsync('rm', ['-rf', path.join(ANDROID_DIR, 'maven/org/webkit/')]);
```
Comment (tsapeta) [inline tools/expotools/src/commands/AndroidBuildPackages.ts]:
```suggestion
  await fs.remove(path.join(ANDROID_DIR, 'maven/org/webkit/'));
```

File: tools/expotools/src/commands/AndroidBuildPackages.ts [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+  await spawnAsync('cp', [
```
Comment (tsapeta) [inline tools/expotools/src/commands/AndroidBuildPackages.ts]: `fs.copy`

File: tools/expotools/src/commands/AndroidBuildPackages.ts [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+    `// FLAG_END_REMOVE__UPDATE_EXPOKIT`,
+    `*/ //`
+  );
+  await _uncommentWhenDistributing([appBuildGradle, expoViewBuildGradle]);
+  await _commentWhenDistributing([constantsJava, expoViewBuildGradle, multipleVersionReactNativeActivity]);
+
+  // Clear maven local so that we don't end up with multiple versions
+  console.log(' ❌  Clearing old package versions...')
+
+  await spawnAsync('rm', [
```
Comment (tsapeta) [inline tools/expotools/src/commands/AndroidBuildPackages.ts]: `fs.remove`

**Review by tsapeta — CHANGES_REQUESTED** [review]
Looks good except a few things, see comments below 👇

File: tools/expotools/src/commands/AndroidBuildPackages.ts (hunk 3 above)
Comment (ide) [reply tools/expotools/src/commands/AndroidBuildPackages.ts]: Node 10 supports the `recursive` option: https://nodejs.org/docs/latest-v10.x/api/fs.html#fs_fs_mkdir_path_options_callback. IIRC we use Node 10.16 in the Expo repo and can rely on this option.

File: tools/expotools/src/commands/AndroidBuildPackages.ts (hunk 9 above)
Comment (ide) [reply tools/expotools/src/commands/AndroidBuildPackages.ts]: fs.remove doesn't implement `-rf`

Comment (tsapeta) [reply tools/expotools/src/commands/AndroidBuildPackages.ts]: it does, see https://github.com/jprichardson/node-fs-extra/blob/master/docs/remove.md

File: tools/expotools/src/commands/AndroidBuildPackages.ts [hunk 10]
```diff
@@ -328,9 +350,23 @@ async function action(options: ActionOptions) {
...
 }
 
+async function _exitHandler(): Promise<void> {
+  if (Object.keys(savedFiles).length) {
+    console.log('Exited early, cleaning up...');
+    await _restoreFilesAsync();
+  }
+}
+
+process.on('SIGINT', _exitHandler);
```
Comment (ide) [inline tools/expotools/src/commands/AndroidBuildPackages.ts]: do you want to handle other signals as well (SIGKILL, SIGTERM) or just SIGINT?

**Review by ide — APPROVED** [review]

File: tools/expotools/src/commands/AndroidBuildPackages.ts (hunk 10 above)
Comment (esamelson) [reply tools/expotools/src/commands/AndroidBuildPackages.ts]: sure, shouldn't hurt anything to handle those cases as well

**Pull request merged**